File: python_solutions/day_7.py
from abc import ABC, abstractmethod
from dataclasses import dataclass
from functools import singledispatchmethod
from pathlib import Path
import re
from typing import Dict, List, Optional

INPUT_FILE_PATH = Path("../inputs/day_7.txt")
CHANGE_DIR_COMMAND = "$ cd"
LIST_CONTENTS_COMMAND = "$ ls"

DISK_SIZE = 70000000
REQUIRED_SPACE = 30000000


class FileSystemObject(ABC):
    def __init__(self, name: str, parent: Optional["Directory"] = None) -> None:
        self._name = name
        self._parent = parent

    @property
    def name(self) -> str:
        return self._name

    @property
    def parent(self) -> Optional["Directory"]:
        return self._parent

    @property
    @abstractmethod
    def size(self) -> int:
        raise NotImplementedError()


class File(FileSystemObject):
    def __init__(
        self, name: str, size: int, parent: Optional["Directory"] = None
    ) -> None:
        super().__init__(name, parent)
        self._size = size

    @property
    def size(self) -> int:
        return self._size


class Directory(FileSystemObject):
    def __init__(self, name: str, parent: Optional["Directory"] = None) -> None:
        super().__init__(name, parent)
        self._contents: Dict[str, FileSystemObject] = {}

    def add_item(self, item: FileSystemObject) -> None:
        self._contents[item.name] = item

    @property
    def contents(self) -> Dict[str, FileSystemObject]:
        return self._contents

    def get_subdirectory(self, name: str) -> "Directory":
        item = self._contents[name]
        if isinstance(item, Directory):
            return item
        else:
            raise ValueError(f"No directory names {name} exists in this directory.")

    @property
    def size(self) -> int:
        return sum([obj.size for obj in self._contents.values()])


def file_system_object_factory(line: str, current_dir: Directory) -> FileSystemObject:
    dir_regex = re.compile(r"dir \w+")
    file_regex = re.compile(r"\d+ \w+")
    if dir_regex.match(line):
        return Directory(name=line[4:], parent=current_dir)
    elif file_regex.match(line):
        return File(name=line.split()[1], size=int(line.split()[0]), parent=current_dir)
    else:
        raise ValueError("Line does not describe a file system object.")


@dataclass
class OSCommand:
    pass


@dataclass
class ListCommand(OSCommand):
    pass


@dataclass
class ChangeDirectoryCommand(OSCommand):
    destination: Directory


class UnrecognisedCommandError(ValueError):
    pass


class CurrentDirectoryNotSetError(RuntimeError):
    pass


class DirectoryHasNoParentError(RuntimeError):
    pass


class ScapingFailedError(RuntimeError):
    pass


def command_factory(line: str, current_dir: Optional[Directory] = None) -> OSCommand:
    if line[: len(CHANGE_DIR_COMMAND)] == CHANGE_DIR_COMMAND:
        dir_name = line[len(CHANGE_DIR_COMMAND) + 1 :]
        if current_dir is None:
            destination = Directory(name=dir_name)
        elif dir_name == "..":
            if current_dir.parent is not None:
                destination = current_dir.parent
            else:
                raise DirectoryHasNoParentError()
        else:
            destination = current_dir.get_subdirectory(dir_name)
        return ChangeDirectoryCommand(destination=destination)
    elif line == LIST_CONTENTS_COMMAND:
        return ListCommand()
    else:
        raise UnrecognisedCommandError()


class FileSystemScraper:
    def __init__(self) -> None:
        self._top_dir: Optional[Directory] = None
        self._current_dir: Optional[Directory] = None

    def scrape(self) -> Directory:
        with open(INPUT_FILE_PATH, "r") as f:
            for line in f.readlines():
                line = line.strip()
                try:
                    command = command_factory(line, self._current_dir)
                    self._handle_command(command)
                except UnrecognisedCommandError:
                    self._handle_file_system_object(line)
        if self._top_dir is not None:
            return self._top_dir
        else:
            raise ScapingFailedError()

    def _handle_file_system_object(self, line: str):
        if self._current_dir is not None:
            self._current_dir.add_item(
                file_system_object_factory(line, self._current_dir)
            )
        else:
            raise CurrentDirectoryNotSetError()

    @singledispatchmethod
    def _handle_command(self, command: OSCommand) -> None:
        raise NotImplementedError()

    @_handle_command.register
    def _(self, command: ChangeDirectoryCommand) -> None:
        if self._top_dir is None:
            self._top_dir = command.destination
        self._current_dir = command.destination

    @_handle_command.register
    def _(self, command: ListCommand) -> None:
        pass


def find_subdirectories(dir: Directory) -> List[Directory]:
    dirs = [dir]
    for item in dir.contents.values():
        if isinstance(item, Directory):
            dirs += find_subdirectories(item)
    return dirs


if __name__ == "__main__":

    scraper = FileSystemScraper()
    top_dir = scraper.scrape()
    print(f"Total file system size: {top_dir.size}")

    dirs = find_subdirectories(top_dir)
    print(f"There are {len(dirs)} directories")

    dir_sizes = [dir.size for dir in dirs]
    small_dirs = [s for s in dir_sizes if s <= 100000]
    print(f"Total size of directories with size <= 100000 is {sum(small_dirs)}")

    current_free_disk_space = DISK_SIZE - top_dir.size
    extra_space_required = REQUIRED_SPACE - current_free_disk_space

    sorted_dir_sizes = sorted(dir_sizes)
    for dir_size in sorted_dir_sizes:
        if dir_size >= extra_space_required:
            break

    print(f"We need {extra_space_required} of extra space.")
    print(f"We can delete a directory that wil provide {dir_size} of extra space.")

File: cpp_solutions/day_7/day_7.cpp
#include <typeinfo>
#include <regex>
#include <iostream>
#include <memory>
#include <fstream>
#include <exception>
#include <algorithm>
#include "day_7.hpp"

#define INPUT_FILE_PATH "../../inputs/day_7.txt"

#define CHANGE_DIR_COMMAND "$ cd"
#define LIST_CONTENTS_COMMAND "$ ls"

#define DISK_SIZE 70000000
#define REQUIRED_SPACE 30000000

FileSystemObject::FileSystemObject(std::string name) : name(name) {};
FileSystemObject::FileSystemObject(
    std::string name, Directory* parent
    ) : name(name), parent(parent) {};

File::File(
    std::string name, int size, Directory* parent
    ) : FileSystemObject(name, parent), size(size) {};

Directory::Directory(std::string name) : FileSystemObject(name) {};
Directory::Directory(
    std::string name, Directory* parent
    ) : FileSystemObject(name, parent) {};

void Directory::addItem(std::unique_ptr<FileSystemObject> item) {
    contents[item->name] = std::move(item);
};

int Directory::getSize() {
    int totalSize = 0;
    for (auto const& [name, item] : contents) {
        totalSize += item->getSize();
    }
    return totalSize;
};

Directory* Directory::getSubDirectory(std::string name) {
    try {
        FileSystemObject* obj = contents.at(name).get();
        if (Directory* dir = dynamic_cast<Directory*>(obj)) {
            return dir;
        };
        throw std::runtime_error("No directory exists with that name.");
    } catch(std::out_of_range) {
        throw std::runtime_error("No item with that name exists in this directory.");
    };
}

std::unique_ptr<FileSystemObject> fileSystemObjectFactory(std::string line, Directory* currentDir) {
    std::regex dirRegex("dir \\w+"), fileRegex("\\d+ \\w+");
    std::smatch dirResult, fileResult;
    std::regex_search(line, dirResult, dirRegex);
    std::regex_search(line, fileResult, fileRegex);
    if (!dirResult.empty()) {
        std::string dirMatch = dirResult[0].str();
        std::string name = dirMatch.substr(4, dirMatch.size());
        return std::make_unique<Directory>(name, currentDir);
    } else if (!fileResult.empty()) {
        std::string dirMatch = fileResult[0].str();
        int delimiterIndex = dirMatch.find(" ");
        int size = stoi(dirMatch.substr(0, delimiterIndex));
        std::string name = dirMatch.substr(delimiterIndex + 1, dirMatch.size());
        return std::make_unique<File>(name, size, currentDir);
    } else {
        throw std::runtime_error("Line does not describe a file system object");
    };
}

Directory* getDirToChangeTo(std::string line, Directory* currentDir) {
    std::string dirCmd = CHANGE_DIR_COMMAND;
    std::string dirName = line.substr(dirCmd.size() + 1, line.size());
    if (dirName == "..") {
        if (currentDir->parent) {
            return currentDir->parent;
        };
        throw std::runtime_error("Directory has no parent");
    } else {
        return currentDir->getSubDirectory(dirName);
    };
}

OSCommand parseCommand(std::string line) {
    std::string dirCmd = CHANGE_DIR_COMMAND;
    std::string lsCmd = LIST_CONTENTS_COMMAND;
    if (line.substr(0, dirCmd.size()) == dirCmd) {
        return OSCommand::CD;
    } else if (line.substr(0, lsCmd.size()) == lsCmd) {
        return OSCommand::LS;
    } else {
        throw std::runtime_error("Provided line is not a valid command.");
    };
};

Directory* FileSystemScraper::scrape() {
    std::ifstream file(INPUT_FILE_PATH);
    std::string line;
    std::getline(file, line);
    while (std::getline(file, line)) {
        if (lineIsCommand(line)) {
            OSCommand command = parseCommand(line);
            handleCommand(line, command);
        } else {
            handleFileSystemObject(line);
        };
    };
    return &topDirectory;
};

void FileSystemScraper::handleFileSystemObject(std::string line) {
    currentDirectory->addItem(fileSystemObjectFactory(line, currentDirectory));
}

void FileSystemScraper::handleCommand(std::string line, OSCommand command) {

    if (command == OSCommand::CD) {
         currentDirectory = getDirToChangeTo(line, currentDirectory);
    } else if (command == OSCommand::LS) {
        // do nothing
    } else {
        throw std::runtime_error("Command not recognised.");
    };
}

std::vector<Directory*> Directory::recursivelyFindSubdirectories() {
    std::vector<Directory*> directories;
    directories.push_back(this);
    for (const auto & [name, item] : contents) {
        const FileSystemObject& objRef = *item;
        if (typeid(objRef) == typeid(Directory)) {
            Directory* dir = dynamic_cast<Directory*>(std::addressof(*item));
            std::vector<Directory*> subDirs = dir->recursivelyFindSubdirectories();
            directories.insert(directories.end(), subDirs.begin(), subDirs.end());
        };
    };
    return directories;
}

int main() {
    FileSystemScraper scraper;
    Directory* fileSystem = scraper.scrape();
    std::cout << "The whole file system has size " << fileSystem->getSize() << std::endl;
    std::vector<Directory*> directories = fileSystem->recursivelyFindSubdirectories();
    std::cout << "There are " << directories.size() << " directories" << std::endl;
    int totalSizeOfSmallDirs = 0;

    std::vector<int> sizes;
    for (Directory* dir : directories) {
        if (dir->getSize() <= 100000) {
            totalSizeOfSmallDirs += dir->getSize();
        }
        sizes.push_back(dir->getSize());
    }
    std::cout << "Total size of directories with size<=100000 is " << totalSizeOfSmallDirs << std::endl;

    int current_free_disk_space = DISK_SIZE - fileSystem->getSize();
    int extra_space_required = REQUIRED_SPACE - current_free_disk_space;

    std::sort(sizes.begin(), sizes.end());
    int deletableDirSize;
    for (int size : sizes) {
        if (size >= extra_space_required) {
            deletableDirSize = size;
            break;
        };
    };

    std::cout << "We need " << extra_space_required << " of extra space" << std::endl;
    std::cout << "We can delete a directory that will provide " << deletableDirSize << " of extra space" << std::endl;
}

